protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,965 @@
+                                    ResponseState::Waiting => {},
+                                    ResponseState::Succeeded => {
+                                        // Given that iterator was not the first to contact the peer
+                                        // it will not be made aware of the closer peers discovered
+                                        // to uphold the S/Kademlia disjoint paths guarantee. See
+                                        // [`ClosestDisjointPeersIter::on_success`] for details.
+                                        iter.on_success(&peer, std::iter::empty());
+                                    },
+                                    ResponseState::Failed => {
+                                        iter.on_failure(&peer);
+                                    },
+                                }
+                            },
+                            None => {
+                                // The iterator is the first to contact this peer.
+                                self.contacted_peers.insert(
+                                    peer.clone().into_owned(),
+                                    PeerState::new(i),
+                                );
+                                return PeersIterState::Waiting(Some(Cow::Owned(peer.into_owned())));
+                            },
+                        }
+                    }
+                    PeersIterState::WaitingAtCapacity => {
+                        match state {
+                            Some(PeersIterState::Waiting(Some(_))) => {
+                                // [`ClosestDisjointPeersIter::next`] returns immediately once a
+                                // [`ClosestPeersIter`] yielded a peer. Thus this state is
+                                // unreachable.
+                                unreachable!();
+                            },
+                            Some(PeersIterState::Waiting(None)) => {}
+                            Some(PeersIterState::WaitingAtCapacity) => {}
+                            Some(PeersIterState::Finished) => {
+                                // `state` is never set to `Finished`.
+                                unreachable!();
+                            },
+                            None => state = Some(PeersIterState::WaitingAtCapacity),
+                        };
+
+                        break;
+                    }
+                    PeersIterState::Finished => break,
+                }
+            }
+        }
+
+        state.unwrap_or(PeersIterState::Finished)
+    }
+
+    /// Only finish the [`ClosestPeersIter`]s that yielded a peer within
+    /// `peers`.
+    ///
+    /// `peers` representing the peers that either returned or put a record.
+    ///
+    /// See [`crate::query::Query::may_finish`] for details.
+    pub fn may_finish<I>(&mut self, peers: I)
+    where
+        I: IntoIterator<Item = PeerId>
```

Comment:
```suggestion
        I: IntoIterator<Item = &PeerId>
```

Reply:
See https://github.com/libp2p/rust-libp2p/pull/1473/files#r438703306.

Reply:
But if this method does not need owned items, it shouldn't require them. The borrowing can be done at the call-site, can it not?